<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    <script>
        // promise周边方法 ： 1.实例方法 （实例化对象的方法） 2.静态方法

        // 一、实例方法  
        // 1.catch： 捕捉错误的； 是  onRejected 的一个别名 。 可以捕捉链式的错误 ；
        // let p = new Promise((resolve,reject)=>{
        //     reject("error");
        //     // resolve("success");
        // })

        // p.then(res=>{

        // }).catch(err=>{
        //     console.log("错误了",err);
        // })

        // p.then(res=>{
        //     // throw new Error("错误111");
        // }).then(res=>{

        // }).catch(err=>{  // 只要链式上有错误 那么都会进入 catch函数；
        //     console.log("有错误了",err);
        // })



        // 2.finally ： 执行完了的时候 调用 改函数 ；
        // 完成 ： 无论成功或者失败都是完成；
        // let p = new Promise((resolve,reject)=>{
        //     setTimeout(() => {
        //         // resolve("success");
        //         reject("错了");
        //     }, 1000);
        // })

        // p.then(res=>{
        //     console.log(res);
        // },err=>{
        //     console.log("失败了",err);
        // }).finally(()=>{
        //     console.log("执行完毕了");
        // })


        // 二、promise的静态方法 ；
        // 1. Promise.resolve:快速创建一个成功状态的promise对象；
        // let p  = new Promise(resolve=>{
        //     resolve("hello");
        // })

        // console.log(p);

        // let p = Promise.resolve("hello");
        // console.log(p);

        // 2.Promise.reject(“结果”): 快速创建一个 失败状态的promise对象 ；
        // let p = Promise.reject("error");
        // console.log(p);

        class MyPromise{
            static resolve(res){
                return new Promise(resolve=>{
                    resolve(res);
                })
            }
            static reject(res){
                return new Promise((resolve,reject)=>{
                    reject(res);
                })
            }
        }

        let p = MyPromise.resolve("hello");
        console.log(p);

        let p2 = MyPromise.reject("error");
        console.log(p2);








    </script>
</body>
</html>